packages/grafana-data/src/types/vector.ts

```diff
@@ -1,35 +1,42 @@
-export interface Vector<T = any> {
-  length: number;
+export abstract class Vector<T = any> {
+  // Default iterator, but can be overridden
+  *[Symbol.iterator](): IterableIterator<T> {
+    for (let i = 0; i < this.length; i++) {
+      yield this.get(i);
+    }
+  }
+
+  abstract length: number;
 
   /**
    * Access the value by index (Like an array)
    */
-  get(index: number): T;
+  abstract get(index: number): T;
```

Comment:
Could maybe look into using ES6 proxies so we could index vectors with `[]`, though I've heard there are some performance considerations with proxies.

Reply:
I think this is a good idea. IIRC the primary shape of Vector APIs is driven by https://arrow.apache.org/docs/js/classes/abstractvector.html to be aligned with Apache Arrow implementation. Is that correct @ryantxu ?